docs/src/cli/deploy-a-program.md

```diff
@@ -297,6 +308,74 @@ the program:
+to offline signer, so that future program upgrades will only be possible with a signature of private key residing on 
+offline machine.
+
+Assuming your program has been deployed (using online machine) and its upgrade authority has been changed to
+offline signer (a signer whose private key is generated and stays on machine **disconnected** from internet), 
+a typical setup would consist of 3 different signers (pairs of keys):
+- online signer (used as fee payer for deploying program buffer, deploying/upgrading program itself)
+- offline signer (serves as authority over program upgrades, protects program upgrades from certain 
+  types of attacks), make sure to have a **secure** backup of this signer
+- program signer (typically, generated when program has been compiled into `.so` file)
+
+The general process is as follows:
+1) (use online machine) create buffer
+2) (use online machine) upgrade buffer authority to offline signer
+3) (optional, use a separate online machine) verify the actual buffer on-chain contents
+4) (use offline machine) build and copy a signature for your intent to upgrade program
+5) (use online machine) use this signature to build and broadcast upgrade transactions on-chain
+
+```bash
+# (1) (use online machine) create buffer
+solana program write-buffer <PROGRAM_FILEPATH>
+
+# (2) (use online machine) upgrade buffer authority to offline signer
+solana program set-buffer-authority <ONLINE_SIGNER_PUB_KEY> --new-buffer-authority <OFFLINE_SIGNER_PUB_KEY>
+```
+
+- note, above we have to issue a separate `solana program set-buffer-authority` command (using online machine) to 
+  change authority to be that of offline signer because we don't have access to offline signer private key to sign 
+  up as buffer authority when running `solana program write-buffer` command using online machine (offline signer 
+  private key is stored on offline machine), and `solana program set-buffer-authority` doesn't impose that requirement.
+
+(3) This is where you'd want to verify that program uploaded to your buffer indeed matches source code.
+The most secure way to do it would be to compile program source code on a different online machine (not the same
+online machine that deploys your program, because it could be compromised) and compare resulting `.so` file with
+program data residing on-chain. See how you can [dump on-chain program into a file](deploy-a-program.md#dumping-a-program-to-a-file)
+to verify it.
+
+After program buffer has been verified and is believed to contain the compiled result of intended source code - we are 
+ready to sign program deploy (to upgrade existing program), fill in all required parameters below (<WHATEVER_VALUE> values).
+
+```bash
+# or (4) (use offline machine) get a signature for your intent to upgrade program
+solana program upgrade --sign-only --fee-payer <ONLINE_SIGNER_PUB_KEY> --program-id <PROGRAM_SIGNER_PUB_KEY> --upgrade-authority <OFFLINE_SIGNER> --buffer <BUFFER_PUB_KEY> --blockhash <VALUE>
```

Comment:
Okay, applied in https://github.com/solana-labs/solana/pull/33860/commits/3f8dd632196d0f0cf1c792d9ed4ff493ee0d8212 (the reason I used PROGRAM_SIGNER terminology is to clarify what PROGRAM_ID actually is, but since it's used everywhere else - it's better be consistent, you are right).